src/bin/lockin.rs

```diff
@@ -137,29 +157,57 @@ const APP: () = {
             c.resources.adcs.1.acquire_buffer(),
         ];
 
-        let dac_samples = [
+        let mut dac_samples = [
             c.resources.dacs.0.acquire_buffer(),
             c.resources.dacs.1.acquire_buffer(),
         ];
 
         let lockin = c.resources.lockin;
         let settings = c.resources.settings;
 
-        let timestamp =
-            c.resources.timestamper.latest_timestamp().unwrap_or(None); // Ignore data from timer capture overflows.
-        let (pll_phase, pll_frequency) = c.resources.pll.update(
-            timestamp.map(|t| t as i32),
-            settings.pll_tc[0],
-            settings.pll_tc[1],
-        );
+        let mut pll_frequency = 0;
+
+        let (sample_phase, sample_frequency) = match settings.lockin_mode {
+            LockinMode::External => {
+                let timestamp =
+                    c.resources.timestamper.latest_timestamp().unwrap_or(None); // Ignore data from timer capture overflows.
+                let (pll_phase, frequency) = c.resources.pll.update(
+                    timestamp.map(|t| t as i32),
+                    settings.pll_tc[0],
+                    settings.pll_tc[1],
+                );
+
+                pll_frequency = frequency;
+
+                let sample_frequency = ((pll_frequency
+                    >> design_parameters::SAMPLE_BUFFER_SIZE_LOG2)
+                    as i32)
+                    .wrapping_mul(settings.lockin_harmonic);
+                let sample_phase = settings.lockin_phase.wrapping_add(
+                    pll_phase.wrapping_mul(settings.lockin_harmonic),
+                );
+
+                (sample_phase, sample_frequency)
+            }
+
+            LockinMode::Internal => {
+                // Reference phase and frequency are known.
+                let pll_phase = 0i32;
+                let pll_frequency =
+                    1i32 << (32 - design_parameters::SAMPLE_BUFFER_SIZE_LOG2);
 
-        let sample_frequency = ((pll_frequency
-            >> design_parameters::SAMPLE_BUFFER_SIZE_LOG2)
-            as i32)
-            .wrapping_mul(settings.lockin_harmonic);
-        let sample_phase = settings
-            .lockin_phase
-            .wrapping_add(pll_phase.wrapping_mul(settings.lockin_harmonic));
+                // Demodulation LO phase offset
+                let phase_offset: i32 = 1 << 30;
+
+                let sample_frequency = (pll_frequency as i32)
+                    .wrapping_mul(settings.lockin_harmonic);
+                let sample_phase = phase_offset.wrapping_add(
+                    pll_phase.wrapping_mul(settings.lockin_harmonic),
+                );
```

Comment:
This should be unified with L182-L188, use `settings.phase_offset` as well, and pulled out of the `match`. But I can do that later.